query-grammar/src/query_grammar.rs

```diff
@@ -374,6 +374,30 @@ pub fn parse_to_ast<'a>() -> impl Parser<&'a str, Output = UserInputAst> {
     spaces()
         .with(optional(ast()).skip(eof()))
         .map(|opt_ast| opt_ast.unwrap_or_else(UserInputAst::empty_query))
+        .map(|ast| rewrite_ast(ast))
+}
+
+/// Removes unnecessary children clauses in AST
+///
+/// Motivated by [issue #1433](https://github.com/quickwit-oss/tantivy/issues/1433)
+fn rewrite_ast(mut input: UserInputAst) -> UserInputAst {
+    if let UserInputAst::Clause(terms) = &mut input {
+        for term in terms {
+            rewrite_ast_clause(term);
+        }
+    }
+    input
+}
+
+fn rewrite_ast_clause(input: &mut (Option<Occur>, UserInputAst)) {
+    match input {
+        (None, UserInputAst::Clause(ref mut clauses)) if clauses.len() == 1 => {
+            let new_term = clauses.pop().unwrap(); // safe because clasues.len() == 1
+            input.0 = new_term.0;
```

Comment:
```suggestion
            *input = new_term;
```

or does that not compile?

Reply:
(Theoretically, even `*input = clauses.pop().unwrap();` but I guess this does not pass the borrow checker before Polonius is merged?)

Reply:
+1

Reply:
> (Theoretically, even `*input = clauses.pop().unwrap();` but I guess this does not pass the borrow checker before Polonius is merged?)

In fact it does. I was pretty sure it won't compile! 🤦